File: backend/app/schemas/pet.py
# ============================================================
# 桌宠 schema（二期 P1）
# ============================================================
from __future__ import annotations

from datetime import datetime
from pydantic import BaseModel, Field


class PetConfigOut(BaseModel):
    id: str
    user_id: str
    enabled: bool = True
    size: str = "medium"
    opacity: float = 1.0
    always_on_top: bool = True
    auto_start: bool = False
    position_x: int = 0
    position_y: int = 0
    current_avatar_id: str | None = None
    state_awareness_enabled: bool = True
    action_switch_interval: int = 30
    show_bubble: bool = True
    bubble_duration: int = 5
    click_interaction: bool = True
    draggable: bool = True
    right_click_menu: bool = True
    tts_enabled: bool = False
    tts_voice: str | None = None
    tts_rate: float = 1.0
    tts_pitch: float = 1.0
    tts_volume: float = 0.8
    speak_scene: str = "remind"
    updated_at: datetime


class UpdatePetConfigRequest(BaseModel):
    enabled: bool | None = None
    size: str | None = None
    opacity: float | None = None
    always_on_top: bool | None = None
    auto_start: bool | None = None
    position_x: int | None = None
    position_y: int | None = None
    current_avatar_id: str | None = None
    state_awareness_enabled: bool | None = None
    action_switch_interval: int | None = None
    show_bubble: bool | None = None
    bubble_duration: int | None = None
    click_interaction: bool | None = None
    draggable: bool | None = None
    right_click_menu: bool | None = None
    tts_enabled: bool | None = None
    tts_voice: str | None = None
    tts_rate: float | None = None
    tts_pitch: float | None = None
    tts_volume: float | None = None
    speak_scene: str | None = None


class UpdatePetStatsRequest(BaseModel):
    """四维陪伴数值的**增量**更新（S6-2）。

    用增量而非绝对值：前端的每次互动（抚摸/喂食/玩耍/休息）只知道
    "该加多少"，若传绝对值就需要前端先持有权威状态 —— 那正是我们要
    消除的 localStorage 旧模式。

    命名说明：饱食度用 `satiety`（饱足感）而不是前端旧代码里的
    `hunger`（饥饿度）。旧代码里 `hunger >= 95` 表示"吃饱了"，
    说明它实际表达的也是饱食度，只是名字取反了，容易读错。
    """

    intimacy: int | None = Field(default=None, ge=-100, le=100)
    satiety: int | None = Field(default=None, ge=-100, le=100)
    mood: int | None = Field(default=None, ge=-100, le=100)
    energy: int | None = Field(default=None, ge=-100, le=100)


class PetAvatarOut(BaseModel):
    id: str
    user_id: str
    name: str
    description: str | None = None
    avatar_type: str = "builtin"
    image_url: str | None = None
    mode: str = "simple"
    eye_position: dict = {}
    mouth_position: dict = {}
    action_frames: dict = {}
    tags: list = []
    is_active: bool = False
    sort_order: int = 0
    created_at: datetime
    updated_at: datetime


class CreatePetAvatarRequest(BaseModel):
    name: str = Field(min_length=1, max_length=50)
    description: str | None = None
    image_url: str | None = None
    mode: str = "simple"
    eye_position: dict = {}
    mouth_position: dict = {}
    action_frames: dict = {}
    tags: list = []


class PetStateOut(BaseModel):
    """桌宠当前状态（状态感知）"""
    time_state: str  # morning/work/noon/afternoon/evening/night/late_night
    time_label: str
    action: str  # 当前动作
    action_label: str
    mood: str  # happy/neutral/caring/encouraging
    mood_label: str
    bubble_text: str | None = None
    has_todo_tasks: bool = False
    todo_count: int = 0
    recent_mood: str | None = None

File: frontend/src/api/pet.ts
// ============================================================
// 桌宠 API（二期 P1）
// 配置 + 形象管理 + 状态感知
// ============================================================
import { http } from './http'

export interface PetConfig {
  id: string
  user_id: string
  enabled: boolean
  size: 'small' | 'medium' | 'large'
  opacity: number
  always_on_top: boolean
  auto_start: boolean
  position_x: number
  position_y: number
  current_avatar_id: string | null
  state_awareness_enabled: boolean
  action_switch_interval: number
  show_bubble: boolean
  bubble_duration: number
  click_interaction: boolean
  draggable: boolean
  right_click_menu: boolean
  tts_enabled: boolean
  tts_voice: string | null
  tts_rate: number
  tts_pitch: number
  tts_volume: number
  speak_scene: 'never' | 'remind' | 'interact' | 'celebrate' | 'always'
  updated_at: string
}

export interface PetAvatar {
  id: string
  user_id: string
  name: string
  description: string | null
  avatar_type: 'builtin' | 'custom'
  image_url: string | null
  mode: 'simple' | 'advanced'
  eye_position: Record<string, unknown>
  mouth_position: Record<string, unknown>
  action_frames: Record<string, string[]>
  tags: string[]
  is_active: boolean
  sort_order: number
  created_at: string
  updated_at: string
}

export interface PetState {
  time_state: string
  time_label: string
  action: string
  action_label: string
  mood: string
  mood_label: string
  bubble_text: string | null
  has_todo_tasks: boolean
  todo_count: number
  recent_mood: string | null
}

/**
 * 桌宠四维陪伴数值（S6-2）。
 *
 * 命名说明：饱食度用 `satiety`（饱足感），不用旧代码里的 `hunger`（饥饿度）。
 * 旧代码写的是 `hunger >= 95 表示吃饱了` —— 说明它表达的其实也是饱食度，
 * 只是名字取反了，非常容易读错。
 */
export interface PetStats {
  intimacy: number  // 亲密度 0-100
  satiety: number   // 饱食度 0-100
  mood: number      // 心情 0-100
  energy: number    // 精力 0-100
  updated_at: string | null
  /** 本次读取是否发生了按时间的衰减（由后端惰性计算） */
  decayed: boolean
}

/** 互动时的**增量**（不是绝对值） */
export type PetStatsDelta = Partial<
  Record<'intimacy' | 'satiety' | 'mood' | 'energy', number>
>

export const petApi = {
  // 配置
  getConfig() {
    return http.get<PetConfig>('/pet/config')
  },
  updateConfig(data: Partial<PetConfig>) {
    return http.patch<PetConfig>('/pet/config', data)
  },

  // 形象
  listAvatars() {
    return http.get<PetAvatar[]>('/pet/avatars')
  },
  createAvatar(data: Partial<PetAvatar>) {
    return http.post<PetAvatar>('/pet/avatars', data)
  },
  updateAvatar(id: string, data: Partial<PetAvatar>) {
    return http.patch<PetAvatar>(`/pet/avatars/${id}`, data)
  },
  deleteAvatar(id: string) {
    return http.delete<void>(`/pet/avatars/${id}`)
  },
  activateAvatar(id: string) {
    return http.post<PetConfig>(`/pet/avatars/${id}/activate`)
  },

  // 状态感知
  getState() {
    return http.get<PetState>('/pet/state')
  },

  // 四维陪伴数值（S6-2：从 localStorage 迁入数据库）
  getStats() {
    return http.get<PetStats>('/pet/stats')
  },
  updateStats(delta: PetStatsDelta) {
    return http.put<PetStats>('/pet/stats', delta)
  },
}
